docs/observability.asciidoc

```diff
@@ -0,0 +1,251 @@
+  };
+----
+
+While the `req` value in `resurrect` will be:
+[source,ts]
+----
+export interface ResurrectEvent {
+  strategy: string;
+  isAlive: boolean;
+  connection: Connection;
+  name: string;
+  request: {
+    id: any;
+  };
+}
+----
+
+=== Correlation id
+Correlating events can be quite hard, especially if there are many events at the same time. The client offers you an automatic (and configurable) system to help you handle this problem.
+[source,js]
+----
+const { Client } = require('@elastic/elasticsearch')
+const client = new Client({ node: 'http://localhost:9200' })
+
+client.on('request', (err, event) => {
+  const { id } = event.meta.request
+  if (err) {
+    console.log({ error: err, reqId: id })
+  }
+})
+
+client.on('response', (err, event) => {
+  const { id } = event.meta.request
+  if (err) {
+    console.log({ error: err, reqId: id })
+  }
+})
+
+client.search({
+  index: 'my-index',
+  body: { foo: 'bar' }
+}, (err, result) => {
+  if (err) console.log(err)
+})
+----
+
+By default the id is an incremental integer, but you can easily configure that with the `generateRequestId` option:
+[source,js]
+----
+const { Client } = require('@elastic/elasticsearch')
+const client = new Client({
+  node: 'http://localhost:9200',
+  generateRequestId: function () {
+    // your id generation logic
+    // must be syncronous
+    return 'id'
+  }
+})
+----
+
+You can also specify a custom id per request:
+[source,js]
+----
+client.search({
+  index: 'my-index',
+  body: { foo: 'bar' }
+}, {
+  id: 'custom-id'
+}, (err, result) => {
+  if (err) console.log(err)
+})
+----
+
+=== Context object
+Sometimes, you might need to make some custom data available in your events, you can do that via the `context` option of a request: 
+[source,js]
+----
+const { Client } = require('@elastic/elasticsearch')
+const client = new Client({ node: 'http://localhost:9200' })
+
+client.on('request', (err, event) => {
+  const { id } = event.meta.request
+  const { context } = event.meta
+  if (err) {
+    console.log({ error: err, reqId: id, context })
+  }
+})
+
+client.on('response', (err, event) => {
+  const { id } = event.meta.request
+  const { winter } = event.meta.context
+  if (err) {
+    console.log({ error: err, reqId: id, winter })
+  }
+})
+
+client.search({
+  index: 'my-index',
+  body: { foo: 'bar' }
+}, {
+  context: { winter: 'is coming' }
+}, (err, result) => {
+  if (err) console.log(err)
+})
+----
+
+=== Client name
+If you are using multiple instances of the client or if you are using multiple child clients _(which is the recommended way to have multiple instances of the client)_, you might need to recognize which client you are using, the `name` options will help you in this regard:
+[source,js]
+----
+const { Client } = require('@elastic/elasticsearch')
+const client = new Client({
+  node: 'http://localhost:9200',
+  name: 'parent-client' // default to 'elasticsearch-js'
+})
+
+const child = client.child({
+  name: 'child-client'
```

Comment:
maybe we should default this to `${parent.name} child`?

Reply:
Initially, I thought about having an automatic name for every child client, such as `elasticsearch-js-child-${count}`. But other than increasing the internal complexity *(you will need a singleton to make sure to avoid duplicates)*, I think that if a user is interested in this feature, they will give a meaningful name.